### PR TITLE
Dmstgcn

### PR DESCRIPTION
**Pull request by Kazeya27**

### CODE REVIEW

**Comment by aptx1231** [comment]
上一个PR没有问题的话，不要关闭，继续修改。这样才能看到上次的问题

File: libcity/data/dataset/dataset_subclass/dmstgcn_dataset.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+                             x_train.shape[0] + x_val.shape[0] + x_test.shape[0]) % self.timeslots
+        if self.normal_external:
+            x_train[..., self.output_dim:] = self.ext_scaler.transform(x_train[..., self.output_dim:])
+            y_train[..., self.output_dim:] = self.ext_scaler.transform(y_train[..., self.output_dim:])
+            x_val[..., self.output_dim:] = self.ext_scaler.transform(x_val[..., self.output_dim:])
+            y_val[..., self.output_dim:] = self.ext_scaler.transform(y_val[..., self.output_dim:])
+            x_test[..., self.output_dim:] = self.ext_scaler.transform(x_test[..., self.output_dim:])
+            y_test[..., self.output_dim:] = self.ext_scaler.transform(y_test[..., self.output_dim:])
+
+        x_trains, y_trains, idx_trains = self.shuffle(x_train, y_train, idx_train)
```
Comment (aptx1231) [inline libcity/data/dataset/dataset_subclass/dmstgcn_dataset.py:55]: 为什么要shuffle？不shuffle回怎么样，做一下实验。。没影响就去掉这个shuffle

File: libcity/data/dataset/dataset_subclass/dmstgcn_dataset.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        return self.train_dataloader, self.eval_dataloader, self.test_dataloader
+
+    def get_data_feature(self):
+        return {"scaler": self.scaler, "adj_mx": self.adj_mx, "ext_dim": self.ext_dim,
+                "num_nodes": self.num_nodes, "feature_dim": self.feature_dim,
+                "output_dim": self.output_dim, "num_batches": self.num_batches,
+                "time_slots": self.timeslots}
+
+    def shuffle(self, x, y, idx):
+        permutation = np.random.permutation(x.shape[0])
```
Comment (aptx1231) [inline libcity/data/dataset/dataset_subclass/dmstgcn_dataset.py:76]: 这个shuffle看起来是在batch-size这个维度做的打乱对吗？如果是的话，感觉没必要，可以去掉。。

File: libcity/model/traffic_speed_prediction/DMSTGCN.py [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+        inputs = batch['X']
+        inputs = inputs.permute(0, 3, 2, 1)
+        in_len = inputs.size(3)
+        if in_len < self.receptive_field:
+            xo = nn.functional.pad(inputs, (self.receptive_field - in_len, 0, 0, 0))
+        else:
+            xo = inputs
+        # xo[:,[0]] means primary feature
+        x = self.start_conv(xo[:, [0]])
+        # xo[:,[1]] means auxiliary feature, witch can be set in the raw_data/dataset_name/config.json file
```
Comment (aptx1231) [inline libcity/model/traffic_speed_prediction/DMSTGCN.py:222]: 进一步写一下，config.json 的data_col参数，应该写两个特征，第0个特征是主特征，第1个是辅助特征。
因此本模型只适合PEMSD4和8数据集。详细点。

File: libcity/data/dataset/dataset_subclass/dmstgcn_dataset.py (hunk 2 above)
Comment (Kazeya27) [reply libcity/data/dataset/dataset_subclass/dmstgcn_dataset.py:76]: 对，这是因为我刚看traffic_state_dataset这个类并没有shuffle操作，而作者代码里有，所以我加上了。但是后面发现最后形成dataloader的时候会做shuffle，所以这个地方确实是多余的。
train_dataloader = DataLoader(dataset=train_dataset, batch_size=batch_size,
                                  num_workers=num_workers, collate_fn=collator,
                                  shuffle=shuffle)

**Review by aptx1231 — APPROVED** [review]

**Comment by aptx1231** [comment]
这个也OK，但你得先解决下冲突才能merge了。

**Comment by aptx1231** [comment]
这个解决方法是这样的，你本地在这个PR的分支，pull master，然后合并下冲突，push上来。不要用github这个解决

**Review by aptx1231 — APPROVED** [review]

File: libcity/model/traffic_speed_prediction/DMSTGCN.py (hunk 3 above)
Comment (Kazeya27) [reply libcity/model/traffic_speed_prediction/DMSTGCN.py:224]: 1

File: libcity/data/dataset/dataset_subclass/dmstgcn_dataset.py (hunk 1 above)
Comment (Kazeya27) [reply libcity/data/dataset/dataset_subclass/dmstgcn_dataset.py]: 1
